Add command line option for separate debug info

jarcherNV opened the pull request:
Add command line arg -separate-debug-info which, if provided, produces both a .spv and a .dbg.spv file. The .dbg.spv file contains full debug info and the .spv file has all debug info stripped out.

csyonghe commented on source/slang/slang-emit.cpp:
```diff
@@ -122,6 +122,7 @@
 #include "slang-vm-bytecode.h"
 
 #include <assert.h>
+#include <unordered_set>
```
no need for std, use `HashSet<uint>` instead.

jarcherNV replied:
Done.

csyonghe commented on source/slang/slang-emit.cpp:
```diff
@@ -2091,22 +2092,103 @@ SlangResult emitSPIRVFromIR(
+    // to check if the instruction number is for a debug instruction as
+    // listed in slang-emit-spirv-ops-debug-info-ext.h
+    static const std::unordered_set<uint32_t> debugExtInstNumbers = {
+        1, 35, 102, 103, 107, 20, 101, 8, 10, 11, 5, 2, 6, 108, 3, 4, 23, 24,
```
Can you use SPIRV enums instead of raw values here?

jarcherNV commented:
/regenerate-cmdline-ref

kaizhangNV commented on source/slang/slang-lower-to-ir.cpp:
```diff
@@ -11959,6 +11971,25 @@ static void ensureAllDeclsRec(IRGenContext* context, Decl* decl)
+    char buf[3]; // 2 hex digits + null terminator
+    for (size_t i = 0; i < size; ++i)
+    {
+        snprintf(buf, sizeof(buf), "%02x", data[i]);
```
why can't we use "sb<<data" syntax?

jarcherNV replied:
Perhaps we can. I'll see if I can clean up this logic in a future version.

jarcherNV replied:
I've played around with this, but I'm not too sure how to use that to give a hexadecimal format.

kaizhangNV replied:
sb << CharUtil::getHexChar(data[i])

this might be working.

not saying what you do is wrong, it's just that we try to avoid calling any c/c++ lib functions through our code base.

jarcherNV replied:
Ahh yeah I missed those. I'll update this.

kaizhangNV commented on source/slang/slang-ir.cpp:
```diff
@@ -3393,20 +3393,40 @@ IRInst* IRBuilder::emitDebugSource(UnownedStringSlice fileName, UnownedStringSli
+}
 IRInst* IRBuilder::emitDebugLine(
     IRInst* source,
+    IRInst* debugBuildIdentifier,
```
As discussed, this should not appear in DebugLine.
`debugBuildIdentifier` should be an separate inst appear at debug info section, which DebugLine will be non-semantics inst interleaved with normal spirv inst.

jarcherNV replied:
Also as discussed, just leaving a note that this is fixed now.

kaizhangNV commented on source/slang/slang-ir-insts.h:
```diff
@@ -3438,6 +3438,13 @@ struct IRDebugSource : IRInst
+{
+    IR_LEAF_ISA(DebugBuildIdentifier)
+    IRInst* getBuildIdentifier() { return getOperand(0); }
+    IRInst* getFlags() { return getOperand(1); }
```
What's the usage of this flag?

jarcherNV replied:
I don't know how we should use this flag, but I think it's part of the spec for the instruction: https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#DebugBuildIdentifier

jarcherNV commented:
Current version is my WIP implementation, which just emits the DebugBuildIdentifier IR instruction by itself without using DebugLine. The IR for this is present through all the IR stages and emitOpDebugBuildIdentifier is called inside slang-emit-spirv.cpp but the instruction seems to not be present in the SPIRV output (although wierdly the string containing the hash is). I will continue debugging this.

jarcherNV commented:
Current version should be ok to review now, but it relies on a change to spirv-tools otherwise the added DebugBuildIdentifier instruction will be optimised out by dead code elimination. I will prepare a fix for this.

jarcherNV commented:
I have prepared a fix to prevent elimination of the new instruction here: https://github.com/KhronosGroup/SPIRV-Tools/pull/6166

kaizhangNV commented on source/slang/slang-emit.cpp:
```diff
@@ -2101,22 +2102,164 @@ SlangResult emitSPIRVFromIR(
+    // to check if the instruction number is for a debug instruction as
+    // listed in slang-emit-spirv-ops-debug-info-ext.h
+    static const uint32_t debugExtInstVals[] = {
+        NonSemanticShaderDebugInfo100DebugCompilationUnit,
```
Is this list matched the strip logic in spirv-opt?
I want to make sure we are doing the same thing as spirv-opt.

jarcherNV replied:
It looks like spirv-opt excludes some more instructions than what I included. I will update this.

jarcherNV replied:
Confirming that I have updated this.

kaizhangNV commented on source/slang/slang-ir-inst-defs.h:
```diff
@@ -1372,6 +1372,7 @@ INST(DebugFunction, DebugFunction, 5, 0)
 INST(DebugInlinedVariable, DebugInlinedVariable, 2, 0)
 INST(DebugScope, DebugScope, 2, 0)
 INST(DebugNoScope, DebugNoScope, 1, 0)
+INST(DebugBuildIdentifier, DebugBuildIdentifier, 2, HOISTABLE)
```
why this is HOISTABLE?
This inst will only be a global inst.

csyonghe reviewed (commented):
We should also consider adding a unit test to look for the debug identifier in the spirv that has the debug info.

jarcherNV commented on source/slang/slang-options.cpp:
```diff
@@ -3062,6 +3066,19 @@ SlangResult OptionsParser::_parse(int argc, char const* const* argv)
+                // The main output SPIRV will have all debug info stripped, and obfuscation
+                // is not needed as this only changes names which will now not be present.
+                m_compileRequest->setDebugInfoLevel(SLANG_DEBUG_INFO_LEVEL_MAXIMAL);
+                linkage->m_optionSet.set(OptionKind::Obfuscate, false);
```
Fixed.

jarcherNV commented:
I have updated I have updated this with the following:
- Updated SPIRV parsing to use a helper class rather than the previous duplicated loops.
- Added a unit test using the Slang API.
- Added a new API function to allow getting the debug data and identifier.

kaizhangNV commented on source/slang/slang-emit.cpp:
```diff
@@ -2104,22 +2105,250 @@ SlangResult emitSPIRVFromIR(
+                SpvWord op = getOperand(i);
+                for (int b = 0; b < 4; ++b)
+                {
+                    char c = (char)((op >> (b * 8)) & 0xFF);
```
can you change this to uint8_t instead?

jarcherNV replied:
I could, but I would need to cast it to a char regardless so that it is appended as a char and not an integer value.

kaizhangNV commented on include/slang.h:
```diff
@@ -4163,9 +4164,29 @@ struct IMetadata : public ISlangCastable
+    This is needed for features such as separate debug compilation which
+    output both base and debug spirv.
+ */
+struct ICompileResult
```
change to class.

jarcherNV commented:
Current version should fix most test failures except tests\language-feature\modules\import-in-include\c.slang which I am still trying to resolve.

jarcherNV commented:
/regenerate-cmdline-ref

jarcherNV commented on source/slang/slang-lower-to-ir.cpp:
```diff
@@ -12006,6 +12025,11 @@ RefPtr<IRModule> generateIRForTranslationUnit(
+        // Build identifier is a hash of the source code and compile options.
+        String buildIdentifier = getBuildIdentifierString(translationUnit->getModule());
+        int buildIdentifierFlags = 0;
+        builder->emitDebugBuildIdentifier(buildIdentifier.getUnownedSlice(), buildIdentifierFlags);
```
This makes sense. I have moved the logic for this to after the linking. Unfortunately there is still a failing test. I have sync'd with Joel about this and while he is not blocked yet, he will be soon. We think it may be best to submit this with the check to only emit the instruction if separate debug info is requested so as not to break anything and allow him to work with the new API, and we create a follow-up issue to unconditionally emit the instruction, if that's ok with you.

csyonghe commented on source/slang/slang-lower-to-ir.cpp:
```diff
@@ -603,6 +603,7 @@ struct IRGenContext
     FunctionDeclBase* funcDecl;
 
     bool includeDebugInfo = false;
+    bool useSeparateDebugInfo = false;
```
this should no longer be needed anymore.

jarcherNV replied:
Fixed.

csyonghe commented on include/slang.h:
```diff
@@ -4163,9 +4164,28 @@ struct IMetadata : public ISlangCastable
+    This is needed for features such as separate debug compilation which
+    output both base and debug spirv.
+ */
+struct ICompileResult : public ISlangCastable
```
Should this be `ISlangUnknown`?

kaizhangNV reviewed (approved):
LGTM now, you might need Yong to take another look.

jkwak-work commented:
This turned out to be a breaking change.
I changed the label accordingly.
